### PR TITLE
Never save passwords for this site: api and storage

### PR DESCRIPTION
**Pull request by CDRussell**
Task/Issue URL: https://app.asana.com/0/0/1206017917054330/f 

### Description
This stack of PRs will offer users a way of saying to _never ask to save a password for this site again_. This particularly PR focusses on:
- The underlying DB and the data APIs for adding sites to the excluded list, resetting the list and getting the count from the list.
- It also includes additional dev settings related to the "never saved" sites

### Steps to test this PR
Most of the testing will happen in the PR above, where the UI-based live. For this PR, just need to verify the DB migration happens ok.

#### Database upgrade
There is a database schema change, so need to check the migration is ok.

- [ ] Install from `develop`
- [ ] Add a password (overflow->Passwords, then add manually using ➕ button)
- [ ] Install from this branch. Open Passwords again, and verify the saved password is still there and correct

#### Fresh install
- [ ] Uninstall and re-install from this branch to get a fresh install. Just verify this launches ok to check the DB creation.

#### [Optional] Test dev settings
- [ ] Visit autofill dev settings (`Settings->Autofill dev settings`) and scroll to the bottom to `Never Saved Sites` section
- [ ] Tap on `Add sample site (fill.dev)`. Verify the number of excluded sites changes from 0 to `1`
- [ ] Tap it multiple times; verify it remains `1` (the same site shouldn't be added multiple times)
- [ ] Tap the `Reset Excluded Sites` button and verify it becomes `0`

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1206071669665237

### CODE REVIEW

**Comment by CDRussell** [comment]
Current dependencies on/for this PR:
* `develop`
  * **PR #3997** <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/3997?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #3998** <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/3998?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/duckduckgo/Android/3998?utm_source=stack-comment).

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/securestorage/SecureStorage.kt [hunk 1]
```diff
@@ -122,6 +122,26 @@ interface SecureStorage {
      * If [canAccessSecureStorage] is false when this is invoked, nothing will be done.
      */
     suspend fun deleteWebsiteLoginDetailsWithCredentials(id: Long)
+
+    /**
+     * Adds a [url] to the list of sites for which we'll never ask to save credentials for.
+     */
+    suspend fun addToNeverSaveList(url: String)
```
Comment (cmonfortep) [inline autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/securestorage/SecureStorage.kt:129]: `url` or `domain`? It seems you want to have domain here?

File: autofill/autofill-store/src/main/java/com/duckduckgo/securestorage/store/db/NeverSavedSitesDao.kt [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    fun insert(neverSavedSiteEntity: NeverSavedSiteEntity): Long
+
+    @Query("delete from never_saved_sites")
+    fun clear()
+
+    @Query("select count(*) from never_saved_sites")
+    fun count(): Flow<Int>
+
+    @Query("select count(*) from never_saved_sites where domain = :url")
+    fun isInNeverSaveList(url: String): Boolean
```
Comment (cmonfortep) [inline autofill/autofill-store/src/main/java/com/duckduckgo/securestorage/store/db/NeverSavedSitesDao.kt:38]: better `domain` than url for clarity?

**Review by cmonfortep — APPROVED** [review]
LGTM

just a few nits on naming

**Comment by CDRussell** [comment]
### Merge activity

* **Dec 21, 5:59 PM**: @@CDRussell merged this pull request with [Graphite](https://app.graphite.dev/github/pr/duckduckgo/Android/3998).

**Pull request merged**